src/fitting.py

```diff
@@ -0,0 +1,169 @@
+            Y_lm_feat.append(temp)
+            lm_order.append((L, M))
+    Y_lm_feat = np.array(Y_lm_feat).T
+    return Y_lm_feat, lm_order
+
+def Y_lms_distribution(theta, phi, L_max, B_lms, only_even_Ls=False):
+    """
+    Computes the probability density given a set of thetas and phis to be
+    sampled, an L_max to generate features and the associated array of coeficients.
+
+    Parameters
+    ------------
+    theta: 1D numpy array
+        array of theta angles ranging from 0 to 2*pi. Used to generate Y_lm features
+    phi: 1D numpy array
+        array of phi angles ranging from 0 to pi. Used to generate Y_lm features.
+        Must be the same size as theta
+    L_max: int
+        maximum L value used to generate a truncated array of Y_lms
+    B_lms: 1d numpy array
+        1d array of coeficients associated with a given real Y_lm
+    only_even_Ls: bool
+        if in the molecular frame, only use even L values to generate features
+
+    Returns
+    --------
+    float
+        probability density scaled by the B_00 coeficient
+    """
+    feats, _ = Y_lm_features(theta, phi, L_max, only_even_Ls)
+    product = feats @ B_lms
+    return abs(product)/B_lms[0]
+
+def fit_Y_lms_binning_least_squares(M_xyz, L_max, numb_bins, only_even_Ls=False):
+    """
+    Least Squares Fitting to get B_lm coeficients
+
+    Parameters
+    ------------
+    M_xyz: n by 3 numpy array
+        observered data points representing the distribution in cartesian coordiantes
+    L_max: int
+        Maximum value of L quantum number determing the number of terms to be used in the fits
+    numb_bins: int
+        number of bins in theta and phi axes
+
+    Returns
+    --------
+    B_lms: 1d array
+        optimized B_lm coeficients used for the probability distribution
+    lm_order: list of tuples
+        first index of each tuple is L and the second is M
+    """
+    assert M_xyz.shape[1] == 3
+    assert numb_bins > 0 and int(numb_bins) == numb_bins
+    M_sph = cart_to_spherical(M_xyz)
+
+    angular_hist, theta_hist, phi_hist = np.histogram2d(M_sph[:, 1], M_sph[:, 2], numb_bins,
+                                                                                    range=[[0, 2*np.pi], [0, np.pi]],
+                                                                                    density=True)
+    #estimate bin by center value
+    theta_hist = (theta_hist[1:] + theta_hist[:-1])/2
+    phi_hist = (phi_hist[1:] + phi_hist[:-1])/2
+    theta_hist, phi_hist = np.meshgrid(theta_hist, phi_hist)
+
+    # vectorize 2d histogram and input angles
+    angular_hist_flat = angular_hist.flatten()
+    theta_hist_flat = theta_hist.flatten()
+    phi_hist_flat = phi_hist.flatten()
+
+    # make ylm feature matrix
+    ylm_features, lm_order = Y_lm_features(theta_hist_flat, phi_hist_flat, L_max, only_even_Ls)
+
+    # solve for coeficients
+    B_lms = np.linalg.solve( ylm_features.T @ ylm_features, ylm_features.T @ angular_hist_flat)
```

Comment:
Do we want to add a ridge regularizing parameter to force the Blm's to be smaller?